Integrate Arthas into Dev Mode

geoand opened the pull request:
[Arthas](https://arthas.aliyun.com/doc/en/index.html) is a Java diagnostic tool
open-sourced by Alibaba middleware team.
It is widely adopted and popular among the developers inside Alibaba.
Arthas helps developers in troubleshooting issues without modifying code or restarting servers.

This Pull Request essentially makes Arthas diagnostics available to a Quarkus application
running in Dev Mode without any user actions being necessary (other than opting into
using Arthas). 
When Arthas is started, its URL is printed in the console so users can easily access it.

Here is what the UI looks like after I issue the `help` command:

![Screenshot from 2021-12-17 17-46-09](https://user-images.githubusercontent.com/4374975/146570798-be0f33e6-2f83-469c-b280-c243fe5b098e.png)


*Future work*

Currently this feature is opt-in via config that must be set when launching dev-mode. It might be interesting to enable this from a button in the Dev UI or an action in the console when the application is already running.

** UPDATE **

The PR now simply logs a message about the exact JBang script needed to launch and attach Arthas to the running Quarkus application.

geoand committed:
Integrate Arthas into Dev Mode

[Arthas](https://arthas.aliyun.com/doc/en/index.html) is a Java diagnostic tool
open-sourced by Alibaba middleware team.
It is widely adopted and popular among the developers inside Alibaba.
Arthas helps developers in troubleshooting issues without modifying code or restarting servers.

This Pull Request essentially makes Arthas diagnostics available to a Quarkus application
running in Dev Mode without any user actions being necessary (other than opting into
using Arthas)

geoand commented:
@loicmathieu you might be interested in this 😉

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/ArthasConfig.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * The directory into which to save the arthas-boot jar if it doesn't exist
+     */
+    @ConfigItem(defaultValue = "${user.home}/.quarkus")
```
@aloubyansky just wanted to check that you were not using this directory for other things that could interfere with this?

loicmathieu replied:
Maybe use temp dir to have it automatically deleted at JVM exit ?

gastaldi replied:
The CLI uses the file `~/.quarkus/config.yaml` for adding new registries

aloubyansky replied:
Right, let's not trash it. Perhaps it should have its own subdirectory at least.

geoand replied:
It does - it's added under `arthas`

maxandersen replied:
definitely not a fan of us downloading it into ~/.quarkus - if we really want to do this I would resolve it as dependency.

geoand replied:
As I mentioned in another comment - the dependency is not enough.

geoand replied:
Scratch that, see my comment below about the all-in-one jar.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:
```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
+    }
+
+    private Path downloadArthas(Path arthasDir, String version) {
+        var downloadURL = "https://arthas.aliyun.com/download/" + version;
```
Can't say I'm excited with us downloading a jar by ourselves from a maybe unstable URL. So first, let's use a Maven Central URL to download this, the artifact is there AFAICS: https://search.maven.org/artifact/com.taobao.arthas/arthas-boot/3.5.5/jar .

Given how it works, I suppose using a dependency is out of the question?

geoand replied:
That's not enough unfortunately. For it to work, we need the entire zip.

loicmathieu replied:
I think using an extension and a dependency should be feasible.
You can then locate the JAR from Maven dependency mechanism and launch it.

geoand replied:
Let me try it again

geoand replied:
Tried it again and it does not work - the entire zip is needed.

maxandersen replied:
https://github.com/alibaba/jbang-catalog/blob/main/jbang-catalog.json#L4

its in maven - its at https://repo1.maven.org/maven2/com/taobao/arthas/arthas-boot/3.5.5/arthas-boot-3.5.5-jar-with-dependencies.jar

and it exists as maven dependency too but you will need to resolve it without transitive dependencies.

geoand replied:
OK, let me try this.

geoand replied:
I am not using this.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:
```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
+
+    private void startArthas(Path arthasJar, Integer httpPort) {
+        log.debug("Attempting to start Arthas");
+        ProcessBuilder processBuilder = new ProcessBuilder("java",
```
I remember some more complicated logic to get the right java binary somewhere in the code base but maybe it's just a bad recollection.

geoand replied:
Right, we do have that. Let me see if I can find it

gastaldi replied:
Is it in `io.quarkus.utilities.JavaBinFinder`?

geoand replied:
Fixed

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:
```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
+
+            }
+            if (arthasProcess.isAlive()) {
+                log.infof("Arthas started and is accessible at http://127.0.0.1:%d", 8563);
```
Maybe use localhost? I'm not sure the IPv4 IP works for someone configured with just IPv6.

geoand replied:
Just pushedthat change :)

gsmet reviewed (changes requested):
I added some comments and questions. The only thing I really want fixed is the download URL.

loicmathieu commented on core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java:
```diff
@@ -210,6 +224,117 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
+
+        try {
+            var arthasProcess = processBuilder.start();
+            try {
+                // attaching to the process takes a while
+                // TODO: we should probably tail the logs instead of waiting
+                Thread.sleep(5_000);
+            } catch (InterruptedException ignored) {
```
You should at least poll a few times each second or something like this.

geoand replied:
Sure, we can do this

loicmathieu reviewed (changes requested):
This is indead a good tool but I didn't use it since a long time !

I'm surprised it's comes OOTB and didn't come from an extension, this will looks very opinionated and make a strong dependency between the core and Arthas.

loicmathieu commented:
I never use the web console, only the command line mode.
Maybe we can integrate the command line mode inside the quarkus dev console so there will not be complicated stuff to wait for the server to be ready. But maybe this would be more difficult.

maxandersen commented:
> I never use the web console, only the command line mode.
> Maybe we can integrate the command line mode inside the quarkus dev console so there will not be complicated stuff to wait for the server to be ready. But maybe this would be more difficult.

having it is devconsole is not really useful imo - then you can't see or interact with your own app.
if you want to use arthos in console mode you would just open another terminal and run it directly.

Thus I would much rather prefer it to just print out the necessary command ...` jbang arthas@alibaba 1234` or `java -jar arthas.jar 1234`

I'm not a fan of quarkus core devmode downloading tools like this...its not what users expect nor something we can easily productize.

geoand commented:
I am not going to put any more effort into this honestly. If someone wants to bring it home, great.

Otherwise 🤷🏼

geoand commented:
> I'm not a fan of quarkus core devmode downloading tools like this...its not what users expect nor something we can easily productize.

Why would we ever productize this?

And in any case, it's an opt-in feature...

geoand committed:
Remove Arthas download and launch capability

We simply log a message with exact JBang script a user
needs to execute in order to get an Arthas terminal

geoand commented:
> Thus I would much rather prefer it to just print out the necessary command ...` jbang arthas@alibaba 1234` or `java -jar arthas.jar 1234`

I pushed a second commit doing only this

geoand commented:
Closing due to lack of interest
